File: lecture_010_graph/L010.java
import java.util.ArrayList;

public class L010 {

    public static class edge {
        int v = 0; // destination
        int w = 0; // weight

        edge(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    public static void main(String[] args) {
        solve();
    }

    public static void solve() {

        int N = 7; // no of vertices
        // float [] f = new float[N];
        ArrayList<edge>[] graph = new ArrayList[N];

        constructGraph(graph);
        display(graph);

        // type.===============
        // basic(graph);

        path(graph);
 
    }

    public static void basic(ArrayList<edge>[] graph) {

        removeEdge(graph, 5, 6);
        display(graph);
        addEdge(graph, 5, 6, 10);
        display(graph);
        removeVtx(graph, 3);
        display(graph);

    }

    public static void path(ArrayList<edge>[] graph) {

        boolean[] vis = new boolean[graph.length];
        // System.out.println(hasPath(graph, 2, 6, vis));
        // System.out.println(allPath(graph, 0, 6, vis, "", 0));
        pair_ ans=allPath_pair(graph,0,6,vis);
        System.out.println(ans.cost+" @ "+ans.way);

    }
    
    public static void constructGraph(ArrayList<edge>[] graph) {

        int N = graph.length;
        for (int i = 0; i < N; i++) {
            graph[i] = new ArrayList<edge>(); // allocate memory to each array value
        }

        addEdge(graph, 0, 1, 10);
        addEdge(graph, 0, 3, 10);
        addEdge(graph, 1, 2, 10);
        addEdge(graph, 2, 3, 10);
        addEdge(graph, 3, 4, 10);
        addEdge(graph, 4, 5, 10);
        addEdge(graph, 4, 6, 10);
        addEdge(graph, 5, 6, 10);

    }

    public static void addEdge(ArrayList<edge>[] graph, int u1, int v1, int w1) {

        // bi-directional graph
        graph[u1].add(new edge(v1, w1));
        graph[v1].add(new edge(u1, w1));
    }

    public static void display(ArrayList<edge>[] graph) {

        int N = graph.length;
        for (int i = 0; i < N; i++) {
            System.out.println();
            System.out.print(i + "->");
            for (edge e : graph[i])
                System.out.print("(" + e.v + "," + e.w + ")");

        }

        System.out.println();

    }

    public static int findEdge(ArrayList<edge>[] graph, int u, int v) { // u ke andar v

        for (int i = 0; i < graph[u].size(); i++) {
            edge e = graph[u].get(i);
            if (e.v == v)
                return i;
        }

        return -1;
    }

    public static void removeEdge(ArrayList<edge>[] graph, int u, int v) { // u ke andar v

        int idx1 = findEdge(graph, u, v);
        if (idx1 != -1)
            graph[u].remove(idx1);

        int idx2 = findEdge(graph, v, u);
        if (idx2 != -1)
            graph[v].remove(idx2);

    }

    public static void removeVtx(ArrayList<edge>[] graph, int vtx) {

        while (graph[vtx].size() != 0) {
            edge e = graph[vtx].get(graph[vtx].size() - 1);
            removeEdge(graph, vtx, e.v);
        }
    }

    public static boolean hasPath(ArrayList<edge>[] graph, int src, int des, boolean[] vis) {
        if (src == des)
            return true;

        boolean res = false;
        vis[src] = true;
        for (edge e : graph[src]) {
            if (!vis[e.v])
                res = res || hasPath(graph, e.v, des, vis);
        }
        vis[src] = false;

        return res;
    }

    public static int allPath(ArrayList<edge>[] graph, int src, int des, boolean[] vis, String ans, int w) {
        if (src == des) {
            System.out.println(ans + des + " @ " + w);
            return 1;
        }

        int count = 0;
        vis[src] = true;

        for (edge e : graph[src]) {
            if (!vis[e.v])
                count += allPath(graph, e.v, des, vis, ans + src + "->", w + e.w);
        }

        vis[src] = false;

        return count;
    }

    public static class pair_ {
        int cost = 0;      // destination
        String way = "";   // weight

        pair_(int v, String w) {
            this.cost = v;
            this.way = w;
        }
    }

    //for maximum path 
    public static pair_ allPath_pair(ArrayList<edge>[] graph, int src, int des, boolean[] vis) {
        if (src == des) {
            return new pair_(0,""+des);
        }

        vis[src] = true;
        pair_ p=new pair_(Integer.MIN_VALUE,"");

        for (edge e : graph[src]) {
            if (!vis[e.v])
            {
                pair_ ret=allPath_pair(graph, e.v, des, vis);
                if(ret.cost > p.cost)
                {
                    p.cost=ret.cost+e.w;
                    p.way=ret.way+"<-"+src;
                }
            }
        }

        vis[src] = false;   

        return p;
    }

}

File: lecture_010_graph/L010.cpp
#include <iostream>
#include <vector>

using namespace std;

class Edge{
    
    public :
    
    int v;//destination
    int w; //weight

    Edge(int v, int w){
      this->v=v;
      this->w=w;
    }
 
};

int N=7;// vertices
vector<vector<Edge>> graph(N,vector<Edge>());

void addEdge(int u,int v,int w){
    
   graph[u].push_back(Edge(v,w));
   graph[v].push_back(Edge(u,w));
    
}

void display(){
    
    cout<<endl;
    for(int i=0; i < graph.size(); i++){
        cout<<i<<" -> ";
        for(int j=0; j < graph[i].size(); j++){
            Edge e=graph[i][j];
            cout<<"("<<e.v<<","<<e.w<<")";
        }
        cout<<endl;
    }

}

int findEdge(int r,int c){   //rth node se cth node  
      
    for(int i=0; i < graph[r].size(); i++){
        Edge e=graph[r][i];
        if(e.v == c){
            return i;
        }
    }
    return -1;              

}

void removeEdge(int u,int v){

    int idx1=findEdge(u,v);
    if(idx1!=-1)
        graph[u].erase(graph[u].begin() + idx1);
   

    int idx2=findEdge(v,u);
    if(idx2!=-1)
        graph[v].erase(graph[v].begin() + idx2);   
   
}

void removeVtx(int vtx){
     
    while(graph[vtx].size()!=0)
    {  
       Edge e=graph[vtx][0];
       removeEdge(vtx, e.v);
    }
    
}

void graphConstruct(){
 
 addEdge(0,1,10);
 addEdge(0,3,10);
 addEdge(1,2,10);
 addEdge(2,3,40);
 addEdge(3,4,2);
 addEdge(4,5,2);
 addEdge(4,6,3);
 addEdge(5,6,18);

 display();
 cout<<endl;
}

bool haspath(int src,int des,vector <bool>& arr){
    if(src==des){
        return true;
    }
    arr[src]=true;
    bool res=false;

    for(int i=0; i < graph[src].size(); i++){
        Edge e=graph[src][i];
        if(!arr[e.v]){
            res=res || haspath(e.v, des, arr);
        }
    }
    arr[src]=false;
    return res;
}

int allPath(int src,int des,vector<bool>& arr,string ans)
{   
    if(src==des){
       cout<<ans<<des<<endl;
       return 1;       
    }

    arr[src]=true;
    int count=0;

    for(int i=0; i < graph[src].size(); i++){
        Edge e=graph[src][i];
        if(!arr[e.v]){
           count+=allPath(e.v, des, arr, ans+to_string(src)+"->");    
        }
    }

    arr[src]=false;
    return count;

}

class ansPair
{
    public:
        int heavy=0;
        int light=1e7;;
        int floor=0;;
        int ceil=1e7;
};

void allSolution(int src, int dest, vector<bool> &vis, int w, string ans, ansPair &pair, int data)
{
    if (src == dest)
    {  

       ans = ans + " "+ to_string(dest);
       cout<<ans<<endl;

       pair.heavy=max(pair.heavy,w);
       pair.light=min(pair.light,w);

       if(w>data)
       {
           pair.ceil=min(pair.ceil,w);
       }

       if(w<data)
       {
           pair.floor=max(pair.floor,w);
       }

       return;
    }

    vis[src] = true;

    for (Edge e : graph[src])
    {
        if (!vis[e.v])
            allSolution(e.v, dest, vis, w + e.w, ans + to_string(src) + " ", pair,data);
    }

    vis[src] = false;
}

void question(){

    // vector <bool>arr(N,false);
    // if(haspath(7,7,arr))
    //    cout<<"found";
    
    // cout<<allPath(0,6,arr,"");

    ansPair obj;
    allSolution(0,6,arr,0,"",obj,30);
    cout<<obj.heavy<<" "<<obj.light<<" "<<" "<<obj.floor<<" "<<obj.ceil<<endl;

}

void basic(){ 
    removeEdge(3,4);    
    display();
    removeVtx(6);
    display(); 
}

void solve(){
   graphConstruct();
   //basic();
   question();
}

int main(){

    solve();
    
}